File: Grind75/clone_graph.py
def clone_graph(node: 'Node') -> 'Node':
    if not node:
        return None
    
    visited = {}

    def dfs(current):
        if current in visited:
            return visited[current]
        
        clone = Node(current.val)
        visited[current] = clone
        
        for neighbor in current.neighbors:
            clone.neighbors.append(dfs(neighbor))
        
        return clone

    return dfs(node)

File: Graphs/cloneGraph.js
function cloneGraph(node) {
  if (!node) return null;

  const map = new Map();

  function dfs(n) {
    if (map.has(n.val)) return map.get(n.val);
    const clone = new _Node(n.val);
    map.set(n.val, clone);
    for (const neighbor of n.neighbors) {
      clone.neighbors.push(dfs(neighbor));
    }
    return clone;
  }

  return dfs(node);
}

// Time: O(nodes + edges)
// Space: O(n)
